### PR TITLE
Group flags and values separated by space

### LINKED ISSUE #6154: spack reorders cflags specified in compilers.yaml making them invalid
```
bash-4.2$ grep cflags ~/.spack/linux/compilers.yaml
      cflags: -O3 -fp-model fast -xMIC-AVX512
bash-4.2$ spack env numactl |grep CFLAGS
CFLAGS=-O3 fast -xMIC-AVX512 -fp-model
bash-4.2$ icc -O3 -fp-model fast -xMIC-AVX512 ~/petsc/src/benchmarks/sizeof.c  && echo sucess!
sucess!
bash-4.2$ icc -O3 fast -xMIC-AVX512 -fp-model ~/petsc/src/benchmarks/sizeof.c && echo sucess!
icc: error #10236: File not found:  'fast'
icc: command line warning #10159: invalid argument for option '-fp-model'
icc: command line error: no files specified; for help type "icc -help"
bash-4.2$ 

```

### PR DESCRIPTION
**Pull request by scheibelp**
@becker33

Fixes #6154

For compiler options which set values using the syntax "-flag value"
(with a space between the flag and the flag's value) the flag and
value were treated as separate and reordered. This updates Spack's
logic to treat the flag and value as a single unit, even if there is
a space. It assumes that only flags are prefixed with "-".

### CODE REVIEW

File: lib/spack/spack/compiler.py [hunk 1]
```diff
@@ -65,6 +65,23 @@ def dumpversion(compiler_path):
...
 
 
+def tokenize_flags(flag_value):
+    tokens = flag_value.split()
+    if not tokens:
+        return []
+    token_iter = iter(tokens)
+    flag = token_iter.next()
+    flags = []
+    for token in token_iter:
```
Comment (tgamblin) [inline lib/spack/spack/compiler.py]: why do you need all the `iter()` stuff?  Why not just `for token in tokens:`?

**Comment by tgamblin** [comment]
Honestly I think we should revisit the flag semantics -- a map is probably not the right thing.  Does this fix #6179 as well as #6154?

**Comment by tgamblin** [comment]
@scheibelp: ok, if @balay can verify that this fixes his issues on Phi, then let's check this in.

I think that we should also figure out a way to add enough determinism to the flag handling that #6179 is fixed (which may mean being stricter about `.satisfies()` for flags), but that can come later.

@balay: does this help you?

**Comment by tgamblin** [comment]
@scheibelp: ok, if @balay can verify that this fixes his issues on Phi, then let's check this in.

I think that we should also figure out a way to add enough determinism to the flag handling that #6179 is fixed (which may mean being stricter about `.satisfies()` for flags), but that can come later.

@balay: does this help you?

**Comment by scheibelp** [comment]
@tgamblin I think @balay mentioned it helped at https://github.com/spack/spack/issues/6154#issuecomment-342365573, although he brought up cases where the overall flag reordering will break some option specifications

**Comment by tgamblin** [comment]
@scheibelp: can you write a couple test cases and a comment for this?

**Comment by scheibelp** [comment]
> can you write a couple test cases and a comment for this?

I can add tests.

I added a comment in `concretize.py` for https://github.com/spack/spack/pull/6169/commits/8d4e153e31197614594e0ac132186e1e493646c2 - is that what you were looking for?

**Comment by tgamblin** [comment]
sorry I should be more clear -- can you add a docstring to the `tokenize_flags()` function?

**Comment by scheibelp** [comment]
It was mentioned that some compiler flags may begin with `+` (or in any case not `-`) so it may be necessary to update this to allow for other symbols. My goal is to check into that tomorrow.

**Comment by tgamblin** [comment]
@scheibelp: I looked through the manual for the major compilers and couldn't find flags that start with anything but `-`, or that take more than one argument.  The only exception was on windows, where flags start with `/`.  So I think we are good on this one.

**Comment by scheibelp** [comment]
@tgamblin OK tests and docstring are now added so if that looks good this is set.

**Pull request merged**